dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -316,27 +316,30 @@ bool WindowBlockInputStream::returnIfCancelledOrKilled()
 
 Block WindowBlockInputStream::readImpl()
 {
-    const auto & stream = children.back();
-    while (!action.input_is_finished)
-    {
-        if (returnIfCancelledOrKilled())
-            return {};
+    // first try to get result without reading from children
+    auto block = action.tryGetOutputBlock();
+    if (block || action.input_is_finished)
```

Comment:
I think it's unnecessary to check `action.input_is_finished` here as it will be check in the wile-loop.